creator-node/src/services/stateMachineManager/monitoring/utils.js

```diff
@@ -0,0 +1,405 @@
+   *  potentialSyncRequests: {Object[]} array of {...nodeUsers, endpoint: {string} endpoint to sync to }
+   * }
+   * @notice this will issue sync to healthy secondary and update replica set away from unhealthy secondary
+   */
+  static async aggregateReconfigAndPotentialSyncOps(
+    nodeUsers,
+    unhealthyPeers,
+    userSecondarySyncMetricsMap,
+    endpointToSPIdMap,
+    thisContentNodeEndpoint
+  ) {
+    // Parallelize calling _aggregateOps on chunks of 500 nodeUsers at a time
+    const nodeUserBatches = _.chunk(
+      nodeUsers,
+      AGGREGATE_RECONFIG_AND_POTENTIAL_SYNC_OPS_BATCH_SIZE
+    )
+    const results = []
+    for (const nodeUserBatch of nodeUserBatches) {
+      const resultBatch = await Promise.allSettled(
+        nodeUserBatch.map((nodeUser) =>
+          StateMonitoringQueueUtils._aggregateOps(
+            nodeUser,
+            unhealthyPeers,
+            userSecondarySyncMetricsMap[nodeUser.wallet],
+            endpointToSPIdMap,
+            thisContentNodeEndpoint
+          )
+        )
+      )
+      results.push(...resultBatch)
+    }
+
+    // Combine each batch's requiredUpdateReplicaSetOps and potentialSyncRequests
+    let requiredUpdateReplicaSetOps = []
+    let potentialSyncRequests = []
+    for (const promiseResult of results) {
+      // Skip and log failed promises
+      const {
+        status: promiseStatus,
+        value: reconfigAndSyncOps,
+        reason: promiseError
+      } = promiseResult
+      if (promiseStatus !== 'fulfilled') {
+        logger.error(
+          `aggregateReconfigAndPotentialSyncOps encountered unexpected failure: ${
+            promiseError.message || promiseError
+          }`
+        )
+        continue
+      }
+
+      // Combine each promise's requiredUpdateReplicaSetOps and potentialSyncRequests
+      const {
+        requiredUpdateReplicaSetOps: requiredUpdateReplicaSetOpsFromPromise,
+        potentialSyncRequests: potentialSyncRequestsFromPromise
+      } = reconfigAndSyncOps
+      requiredUpdateReplicaSetOps = requiredUpdateReplicaSetOps.concat(
+        requiredUpdateReplicaSetOpsFromPromise
+      )
+      potentialSyncRequests = potentialSyncRequests.concat(
+        potentialSyncRequestsFromPromise
+      )
+    }
+
+    return { requiredUpdateReplicaSetOps, potentialSyncRequests }
+  }
+
+  /**
+   * Used to determine the `requiredUpdateReplicaSetOps` and `potentialSyncRequests` for a given nodeUser.
+   * @param {Object} nodeUser { primary, secondary1, secondary2, primarySpID, secondary1SpID, secondary2SpID, user_id, wallet}
+   * @param {Set<string>} unhealthyPeers set of unhealthy peers
+   * @param {string (secondary endpoint): Object{ successRate: number (0-1), successCount: number, failureCount: number }} userSecondarySyncMetrics mapping of each secondary to the success metrics the nodeUser has had syncing to it
+   * @param {Object} endpointToSPIdMap
+   */
+  static async _aggregateOps(
```

Comment:
i dont love this being a util since it's kind of the main logic in the queue

Reply:
making a note to myself to look into the best approach for this after I add the reconciliation queue and make aggregateOps easier to follow